File: backend/src/pipeline/orchestrator.ts
/**
 * Pipeline orchestrator: drives an application through extract -> structure ->
 * certificates -> evaluate. Fail-fast: any stage error marks the application
 * `failed` (recording which stage broke) and stops; later stages do not run.
 *
 * The heavy lifting is deterministic (the parser's `structured` output). AI is
 * used only when needed: an LLM structuring fallback when parsing is weak, gated
 * vision for certificate images, and the single final evaluation.
 *
 * All side effects go through injected collaborators (`PipelineRepo`, `extract`,
 * `call`, `loadImage`) so the control flow is unit-testable without a DB or network.
 */
import { config } from "../config.js";
import { structureResume, classifyCertificates } from "./submodels.js";
import type { GeminiCaller, ImageLoader, RawImage, ClassifiedImage } from "./submodels.js";
import { candidateFromParser, candidateFromLlm } from "./candidate.js";
import type { Candidate, ExtractionResult } from "./candidate.js";
import { evaluate, type EvaluationResult } from "./evaluator.js";
import type { BasicDetails } from "../db/models/Application.js";

export type { ExtractionResult } from "./candidate.js";

export interface PipelineRepo {
  getApplication(id: string): Promise<{ resumePath: string; jobDescription: string }>;
  setStatus(id: string, status: string, opts?: { errorStage?: string; errorMessage?: string }): Promise<void>;
  saveBasicDetails(id: string, basicDetails: BasicDetails): Promise<void>;
  startRun(id: string, stage: Stage): Promise<void>;
  finishRun(id: string, stage: Stage, output: unknown): Promise<void>;
  failRun(id: string, stage: Stage, error: string): Promise<void>;
  saveExtractedImages(id: string, images: RawImage[]): Promise<void>;
  updateImageClassifications(id: string, classified: ClassifiedImage[]): Promise<void>;
  saveEvaluation(id: string, evaluation: EvaluationResult): Promise<void>;
}

export type Stage = "extract" | "structure" | "certificates" | "evaluate";

export interface ProcessDeps {
  repo: PipelineRepo;
  extract: (appId: string, resumePath: string) => Promise<ExtractionResult>;
  call: GeminiCaller;
  loadImage: ImageLoader;
}

class StageError extends Error {
  constructor(
    public readonly stage: Stage,
    public readonly cause: unknown,
  ) {
    super(cause instanceof Error ? cause.message : String(cause));
    this.name = "StageError";
  }
}

async function runStage<T>(repo: PipelineRepo, id: string, stage: Stage, fn: () => Promise<T>): Promise<T> {
  await repo.startRun(id, stage);
  try {
    const result = await fn();
    await repo.finishRun(id, stage, result);
    return result;
  } catch (err) {
    await repo.failRun(id, stage, err instanceof Error ? err.message : String(err));
    throw new StageError(stage, err);
  }
}

export async function processApplication(id: string, deps: ProcessDeps): Promise<void> {
  const { repo, extract, call, loadImage } = deps;
  const app = await repo.getApplication(id);

  try {
    await repo.setStatus(id, "processing");

    const raw = await runStage(repo, id, "extract", () => extract(id, app.resumePath));
    // Persist the cheap, LLM-free basics + images BEFORE any Gemini call, so a
    // later stage failure still leaves the dashboard with usable details.
    await repo.saveBasicDetails(id, raw.basic_details);
    await repo.saveExtractedImages(id, raw.pipeline_b.images);

    // STRUCTURE: trust the parser's deterministic output; only fall back to the
    // LLM when parsing came out weak (messy/unusual layout).
    const candidate: Candidate = await runStage(repo, id, "structure", async () => {

      if (raw.parse_quality.status === "good") {
        return candidateFromParser(raw.structured, raw.links);
      }
      const llmResume = await structureResume(raw.pipeline_a, call);
      return candidateFromLlm(llmResume, raw.links);
    });

    // CERTIFICATES: gated + capped vision. Icons/logos are skipped (their category
    // already comes from the link domain); only certificate-like images reach the LLM.
    const certImages = raw.pipeline_b.images
      .filter((img) => img.likely_certificate)
      .slice(0, config.maxVisionImages);
    const certs = await runStage(repo, id, "certificates", () =>
      classifyCertificates(certImages, call, loadImage),
    );
    await repo.updateImageClassifications(id, certs);

    // EVALUATE: the single guaranteed LLM call, over the compact candidate JSON.
    const evaluation = await runStage(repo, id, "evaluate", () =>
      evaluate(app.jobDescription, candidate, call),
    );
    await repo.saveEvaluation(id, evaluation);

    await repo.setStatus(id, "completed");
  } catch (err) {
    const stage = err instanceof StageError ? err.stage : "unknown";
    const message = err instanceof Error ? err.message : String(err);
    await repo.setStatus(id, "failed", { errorStage: stage, errorMessage: message });
    throw err;
  }
}

File: backend-py/app/pipeline/orchestrator.py
"""Pipeline orchestrator (port of orchestrator.ts).

Drives an application through extract -> structure -> certificates -> evaluate.
Fail-fast: any stage error marks the application `failed` (recording which stage
broke) and stops; later stages do not run.

All side effects go through injected collaborators (repo, extract, call, load_image)
so the control flow stays isolated from the DB and network.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Protocol

from app.config import settings
from app.pipeline.candidate import candidate_from_llm, candidate_from_parser
from app.pipeline.evaluator import evaluate
from app.pipeline.submodels import classify_certificates, structure_resume

# extract / structure / certificates / evaluate
Stage = str


class PipelineRepo(Protocol):
    async def get_application(self, id: str) -> dict: ...
    async def set_status(
        self, id: str, status: str, error_stage: str | None = None, error_message: str | None = None
    ) -> None: ...
    async def save_basic_details(self, id: str, basic_details: Any) -> None: ...
    async def start_run(self, id: str, stage: Stage) -> None: ...
    async def finish_run(self, id: str, stage: Stage, output: Any) -> None: ...
    async def fail_run(self, id: str, stage: Stage, error: str) -> None: ...
    async def save_extracted_images(self, id: str, images: list[dict]) -> None: ...
    async def update_image_classifications(self, id: str, classified: list[dict]) -> None: ...
    async def save_evaluation(self, id: str, evaluation: dict) -> None: ...


@dataclass
class ProcessDeps:
    repo: PipelineRepo
    extract: Callable[[str, str], Awaitable[dict]]
    call: Callable[..., Awaitable[Any]]
    load_image: Callable[[str], Awaitable[Any]]


class StageError(Exception):
    def __init__(self, stage: Stage, cause: Exception) -> None:
        self.stage = stage
        self.cause = cause
        super().__init__(str(cause))


async def _run_stage(
    repo: PipelineRepo, id: str, stage: Stage, fn: Callable[[], Awaitable[Any]]
) -> Any:
    await repo.start_run(id, stage)
    try:
        result = await fn()
        await repo.finish_run(id, stage, result)
        return result
    except Exception as err:  # noqa: BLE001
        await repo.fail_run(id, stage, str(err))
        raise StageError(stage, err)


async def process_application(id: str, deps: ProcessDeps) -> None:
    repo, extract, call, load_image = deps.repo, deps.extract, deps.call, deps.load_image
    app = await repo.get_application(id)

    try:
        await repo.set_status(id, "processing")

        raw = await _run_stage(repo, id, "extract", lambda: extract(id, app["resumePath"]))
        # Persist the cheap, LLM-free basics + images BEFORE any Gemini call, so a
        # later stage failure still leaves the dashboard with usable details.
        await repo.save_basic_details(id, raw["basic_details"])
        await repo.save_extracted_images(id, raw["pipeline_b"]["images"])

        # STRUCTURE: trust the parser's deterministic output; only fall back to the
        # LLM when parsing came out weak (messy/unusual layout).
        async def do_structure() -> dict:
            if raw["parse_quality"]["status"] == "good":
                return candidate_from_parser(raw["structured"], raw["links"])
            llm_resume = await structure_resume(raw["pipeline_a"], call)
            return candidate_from_llm(llm_resume, raw["links"])

        candidate = await _run_stage(repo, id, "structure", do_structure)

        # CERTIFICATES: gated + capped vision. Icons/logos are skipped; only
        # certificate-like images reach the LLM.
        cert_images = [
            img for img in raw["pipeline_b"]["images"] if img.get("likely_certificate")
        ][: settings.max_vision_images]
        certs = await _run_stage(
            repo, id, "certificates", lambda: classify_certificates(cert_images, call, load_image)
        )
        await repo.update_image_classifications(id, certs)

        # EVALUATE: the single guaranteed LLM call, over the compact candidate JSON.
        evaluation = await _run_stage(
            repo, id, "evaluate", lambda: evaluate(app["jobDescription"], candidate, call)
        )
        await repo.save_evaluation(id, evaluation)

        await repo.set_status(id, "completed")
    except Exception as err:  # noqa: BLE001
        stage = err.stage if isinstance(err, StageError) else "unknown"
        message = str(err)
        await repo.set_status(id, "failed", error_stage=stage, error_message=message)
        raise
